File: tools/editor/shared/vfx_harassment.py
"""群体近身侵扰数值闸门：工作台与全工程校验共用。"""
import math


def harassment_errors(value):
    if not isinstance(value, dict):
        return ["harassment 须为对象"]
    errors = []
    for key in ("radius", "height", "attackPerSecond"):
        v = value.get(key)
        if (not isinstance(v, (int, float)) or isinstance(v, bool) or not math.isfinite(v)
                or v < 0 or (key == "radius" and v == 0)):
            errors.append(f"harassment.{key} 须为{'正' if key == 'radius' else '非负'}有限数")
    return errors

File: src/systems/vfx/vfxHarassment.ts
import type { VfxEmitterRuntime } from './vfxSim';
import type { Vec3 } from '../../utils/sceneSpace';

/** 读模拟的真实个体；不以巢/发射器原点代替接触，不把视觉数量当攻击倍率。 */
export function flockHarassmentRate(emitter: VfxEmitterRuntime, player: Vec3): number {
  const h = emitter.def.behavior?.harassment;
  if (!h || !emitter.active || emitter.flock?.state !== 'airborne') return 0;
  if (![h.radius, h.height, h.attackPerSecond].every(Number.isFinite)
    || h.radius <= 0 || h.height < 0 || h.attackPerSecond <= 0) return 0;
  const p = emitter.p;
  for (let i = 0; i < p.cap; i++) {
    if (!p.alive[i]) continue;
    const dx = p.x[i] - player[0], dy = p.y[i] - player[1] - h.height, dz = p.z[i] - player[2];
    if (dx * dx + dy * dy + dz * dz <= h.radius * h.radius) return h.attackPerSecond;
  }
  return 0;
}
